lib/charms/grafana_k8s/v1/grafana_source.py

```diff
@@ -0,0 +1,336 @@
+    ) -> None:
+        """Construct a Grafana charm client.
+
+        The :class:`GrafanaSourceConsumer` object provides an interface
+        to Grafana. This interface supports providing additional
+        sources for Grafana to monitor. For example, if a charm
+        exposes some metrics which are consumable by an ingestor
+        (such as Prometheus), then an additional source can be added
+        by instantiating a :class:`GrafanaSourceConsumer` object and
+        adding its datasources as follows:
+
+            self.grafana = GrafanaSourceConsumer(
+                self, "grafana-datasource", {"grafana_datasource"}: ">=2.0"}
+            )
+            self.grafana.add_source(
+                address=<address>,
+                port=<port>
+            )
+
+        Args:
+
+            charm: a :class:`CharmBase` object which manages this
+                :class:`GrafanaSourceConsumer` object. Generally this is
+                `self` in the instantiating class.
+            name: a :string: name of the relation between `charm`
+                the Grafana charmed service.
+            consumes: a :dict: of acceptable monitoring service
+                providers. The keys of the dictionary are :string:
+                names of grafana source service providers. Typically,
+                this is `grafana-source`. The values of the dictionary
+                are corresponding minimal acceptable semantic versions
+                for the service.
+            multi: an optional (default `False`) flag to indicate if
+                this object should support interacting with multiple
+                service providers.
+
+        """
+        super().__init__(charm, name, consumes, multi)
+
+        self.charm = charm
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
+        events = self.charm.on[name]
+        self.framework.observe(events.relation_joined, self._set_sources)
+
+    def add_source(self, address: str, port: str,
+                   source_type: str = "prometheus", source_name: Optional[str] = None,
+                   rel_id: Optional[int] = None) -> None:
+        """Add an additional source to the Grafana source service.
+
+        Args:
+            address: a :str: which gives address of the target
+            port: a :str: which gives the port of the target
+            source_type: a :str: indicating the source type. Defaults to "prometheus"
+            source_name: a :str: specifying the source name. Otherwise this is calculated
+            rel_id: an optional integer specifying the relation ID
+                for the grafana source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+        rel_id = rel_id if rel_id is not None else rel.id
```

Comment:
If the caller did not provide `rel_id`, this will work only if there is only one relation in place.
Instead, shouldn't this iterate over `self.model.relations`?

Reply:
This will throw an exception from the framework if possible. But this is a consumer library. How many `grafana-source` relations do you think it's going to have?

If it needs more than one, that will be in the Charm, and the author will see exceptions/know to pass the ID specific to that relation.

Reply:
an innocent user experimenting with two grafanas at the same time.
nobody wants to touch relation ids, even with a pole.